src/lib/staging-environment/StagingEnvironmentFactory.ts

```diff
@@ -0,0 +1,101 @@
+import Axios, { AxiosInstance } from 'axios';
+import { encodeQueryString } from '../utils/Url';
+import { StagingEnvironmentFactoryConfig } from './StagingEnvironmentFactoryConfig';
+import { GenerateDomainOptions } from './GenerateDomainOptions';
+
+/**
+ * Default StagingEnvironmentFactoryConfig values
+ */
+const DefaultStagingEnvironmentFactoryConfig: StagingEnvironmentFactoryConfig = {
+  baseUrl: 'https://virtual-staging.adis.ws'
+};
+
+/**
+ * Amplience [Virtual Staging Environment](https://docs.amplience.net/virtualstaging/virtualstagingintro.html?h=vse) factory.
+ *
+ * This factory is intended to be used by end user applications to preview content for a given snapshotID or at a given timestamp.
+ *
+ * It works by passing the virtual staging environment domain and the snapshotID or timestamp to the virtual-staging API,
+ * which will return a new domain that can be used in the ContentClient.
+ *
+ * Example:
+ *
+ * ```typescript
+ * const factory = new StagingEnvironmentFactory('11122222gggg232gg32g32g.staging.bigcontent.io');
+ * const stagingEnvironmentWithSnapshot = await factory.generateDomain({snapshotId: 'abcdef123456'});
+ *
+ * const client = new ContentClient({
+ *   account: 'test',
+ *   stagingEnvironment: stagingEnvironmentWithSnapshot
+ * });
+ *
+ * ```
+ *
+ * You may override other settings when constructing the client but if no additional configuration is provided sensible defaults will be used.
+ */
+export class StagingEnvironmentFactory {
+  private readonly client: AxiosInstance;
+
+  /**
+   * Creates a StagingEnvironmentFactory instance.
+   * @param string stagingEnvironment
+   * @param config Client configuration options
+   */
+  constructor(
+    private readonly stagingEnvironment: string,
+    config: StagingEnvironmentFactoryConfig = {}
+  ) {
+    this.client = this.createClient({
+      ...DefaultStagingEnvironmentFactoryConfig,
+      ...config
+    });
+  }
+
+  /**
+   * Generates a new staging environment domain name using the snapshotId and/or timestamp thats supplied in the options argument
+   * @param options
+   */
+  generateDomain(options: GenerateDomainOptions): Promise<string> {
+    const url = this.buildUrl(options);
+
+    return this.client
+      .get<string>(url)
+      .then(response => {
+        return Promise.resolve(response.data);
```

Comment:
You don't need to return a promise in a then it already returns a promise. also best to use async await :)